'PTHREAD_MUTEX_ERRORCHECK_NP' undeclared

Hi, i'm trying to build ocaml-multicore on alpine linux, which relies 
on musl libc rather than the glibc. 

```

# gcc -c -O2 -fno-strict-aliasing -fwrapv -Wall -fno-tree-vrp  -g -D_FILE$
# gcc -c -O2 -fno-strict-aliasing -fwrapv -Wall -fno-tree-vrp  -g -D_FILE$
# gcc -c -O2 -fno-strict-aliasing -fwrapv -Wall -fno-tree-vrp  -g -D_FILE$
# platform.c: In function 'caml_plat_mutex_init':
# platform.c:25:34: error: 'PTHREAD_MUTEX_ERRORCHECK_NP' undeclared (firs$
#    pthread_mutexattr_settype(&ma, PTHREAD_MUTEX_ERRORCHECK_NP);
#                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~
#                                   PTHREAD_MUTEX_ERRORCHECK

```
Understandably gcc complains it cannot find the non-portable version
`PTHREAD_MUTEX_ERRORCHECK_NP`

I understand what `PTHREAD_MUTEX_ERRORCHECK` does ([Posix manual](http://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html))

But I am not sure about `PTHREAD_MUTEX_ERRORCHECK_NP`.
I Couldn't find much information... 
I tried to have a look into glibc source, 
found something about not doing lock elision, ...

Could someone help me understand what `PTHREAD_MUTEX_ERRORCHECK_NP` does?

Also, in `platform.c`, should I simply discard 
`#define _GNU_SOURCE /* for PTHREAD_MUTEX_ERRORCHECK_NP */`
and use the standard `PTHREAD_MUTEX_ERRORCHECK` ?

I was able to build successfully doing so but I suspect the `_NP` was here for a reason...

--- a/runtime/caml/platform.h
+++ b/runtime/caml/platform.h
@@ -2,9 +2,6 @@
 #define CAML_PLAT_THREADS_H
 /* Platform-specific concurrency and memory primitives */
 
-#ifdef __linux__
-#define _GNU_SOURCE /* for PTHREAD_ERRORCHECK_MUTEX_INITIALIZER_NP */
-#endif
 #include <pthread.h>
 #include <errno.h>
 #include <string.h>

--- a/runtime/platform.c
+++ b/runtime/platform.c
@@ -1,6 +1,5 @@
 #define CAML_INTERNALS
 
-#define _GNU_SOURCE /* for PTHREAD_MUTEX_ERRORCHECK_NP */
 #include <sys/mman.h>
 #include <string.h>
 #include <unistd.h>
@@ -11,16 +10,20 @@
 
 /* Mutexes */
 
-void caml_plat_mutex_init(caml_plat_mutex* m)
+void caml_plat_mutex_init(caml_plat_mutex * m)
 {
-#ifdef __linux__
-  pthread_mutexattr_t ma;
-  pthread_mutexattr_init(&ma);
-  pthread_mutexattr_settype(&ma, PTHREAD_MUTEX_ERRORCHECK_NP);
-  check_err("mutex_init", pthread_mutex_init(m, &ma));
-#else
-  check_err("mutex_init", pthread_mutex_init(m, 0));
-#endif
+  int rc;
+  pthread_mutexattr_t attr;
+  rc = pthread_mutexattr_init(&attr);
+  if (rc != 0) goto error1;
+  rc = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
+  if (rc != 0) goto error2;
+  rc = pthread_mutex_init(m, &attr);
+  // fall through
+error2:
+  pthread_mutexattr_destroy(&attr);
+error1:
+  check_err("mutex_init", rc);
 }
 
 void caml_plat_assert_locked(caml_plat_mutex* m)

